File: libasynDiscordWebHook-PM/src/ialdrich23xx/libasynwebhook/discord/body/embed/components/Timestamp.php
<?php

declare(strict_types=1);

namespace ialdrich23xx\libasynwebhook\discord\body\embed\components;

use DateTime;
use DateTimeZone;
use ialdrich23xx\libasynwebhook\discord\body\embed\base\Structure;
use ialdrich23xx\libasynwebhook\Loader;
use function date_default_timezone_get;
use function in_array;
use function timezone_identifiers_list;

class Timestamp extends Structure
{
    const FORMAT = "Y-m-d\TH:i:s.v\Z";

    private string $timezone;

    public function __construct(
        private DateTime $data
    ) {
        $this->timezone = date_default_timezone_get();
    }

    public static function make(DateTime $data): self
    {
        return new self($data);
    }

    public function getData(): DateTime
    {
        return $this->data;
    }

    public function setTimeZone(string $timeZone): self
    {
        $this->timezone = $timeZone;

        if (!$this->isValidTimeZone()) {
            Loader::getInstance()->getPlugin()->getLogger()->error($timeZone . "not is timezone");

            $this->timezone = date_default_timezone_get();
        }

        return $this;
    }

    public function isValidTimeZone(): bool
    {
        return in_array($this->timezone, timezone_identifiers_list());
    }

    public function build(): bool
    {
        if (!$this->isValidTimeZone()) return false;

        $this->data->setTimezone(new DateTimeZone($this->timezone));

        return true;
    }

    public function toString(): string
    {
        return "Timestamp(data=" . $this->getData()->format(Timestamp::FORMAT) . ",timezone=" . $this->timezone . ")";
    }

    /**
     * @no-require
     */
    public function toArray(): array
    {
        return [];
    }
}

File: libasynDiscordWebHook-API/src/main/java/com/ialdrich23xx/libasynwebhook/api/discord/body/embed/components/Timestamp.java
package com.ialdrich23xx.libasynwebhook.api.discord.body.embed.components;

import com.ialdrich23xx.libasynwebhook.api.discord.body.embed.base.Structure;

import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

public class Timestamp extends Structure {

    private DateTimeFormatter date;
    private OffsetDateTime format;

    public Timestamp(DateTimeFormatter date, OffsetDateTime format) {
        this.date = date;
        this.format = format;
    }

    public static Timestamp make(DateTimeFormatter date, OffsetDateTime format) {
        return new Timestamp(date, format);
    }

    public static Timestamp make(DateTimeFormatter date) {
        return new Timestamp(date, OffsetDateTime.now(ZoneId.of("UTC")));
    }

    public Timestamp setFormat(OffsetDateTime format) {
        this.format = format;

        return this;
    }

    public OffsetDateTime getFormat() {
        return this.format;
    }

    public DateTimeFormatter getDate() {
        return this.date;
    }

    @Override
    public Boolean build() {
        return date != null;
    }

    /**
     * @no-require
     */
    @Override
    public Map<String, Object> toArray() {
        return new HashMap<>();
    }

    @Override
    public String toString() {
        return "Timestamp(data=" + this.getFormat().format(this.getDate()) + ",timezone=" + this.getFormat().toString() + ")";
    }
}
